src/ObjCRuntime/DisposableObject.cs

```diff
@@ -0,0 +1,95 @@
+//
+// DisposableObject.cs: A base class for for many native data types without assuming any particular lifecycle.
+
+// Authors:
+//   Rolf Bjarne Kvinge
+//
+// Copyright 2021 Microsoft Corp
+//
+using System;
+using System.Runtime.InteropServices;
+using ObjCRuntime;
+using Foundation;
+
+#nullable enable
+
+namespace ObjCRuntime {
+	//
+	// The DisposableObject class is intended to be a base class for many native data
+	// data types, without assuming any particular lifecycle (might be reference counted,
+	// might not be).
+	//
+	// It provides the common boilerplate for this kind of objects and the Dispose
+	// pattern.
+	//
+	public abstract class DisposableObject : INativeObject, IDisposable {
+		IntPtr handle;
+		readonly bool owns;
```

Comment:
This looks like SafeHandle :D

Reply:
> The `owns` parameter seems to put the burden on the subclasses of DisposableObject.

Not sure I understand what you mean here?

Reply:
I'm basing the domain knowledge on SafeHandle and its inner workings, because it seems to have almost the same purpose: https://source.dot.net/#System.Private.CoreLib/SafeHandle.cs,53

From my understanding, the base subclass should be responsable for the last object release and the initial registration for dispose/finalize.

The subclasses shouldn't care about `owns`, as it should automatically be tracked in the underlying DisposableObject - regardless of the tracking mechanism.